MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java

```diff
@@ -0,0 +1,268 @@
+
+            // Make EGL surface current
+            if (!EGL14.eglMakeCurrent(eglDisplay, eglSurface, eglSurface, eglContext)) {
+                glError++;
+                Log.e(TAG, "Failed to make context current");
+            }
+
+            // Check for any errors
+            glError += GLES20.glGetError();
+        });
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
+    public void release() {
+        Log.d(TAG, "Releasing output surface");
+        renderHandler.post(() -> {
+            if (eglDisplay != EGL14.EGL_NO_DISPLAY) {
+                EGL14.eglDestroySurface(eglDisplay, eglSurface);
+                EGL14.eglDestroyContext(eglDisplay, eglContext);
+                EGL14.eglReleaseThread();
+                EGL14.eglTerminate(eglDisplay);
+            }
+
+            eglDisplay = EGL14.EGL_NO_DISPLAY;
+            eglContext = EGL14.EGL_NO_CONTEXT;
+            eglSurface = EGL14.EGL_NO_SURFACE;
+
+            textureRender = null;
+            surfaceTexture.release();
+            surfaceTexture = null;
+            decodeSurface.release();
+            decodeSurface = null;
+
+            // Check for any errors
+            glError += GLES20.glGetError();
+        });
+    }
+
+    public Surface getDecodeSurface() {
+        SettableFuture<Surface> surfaceFuture = SettableFuture.create();
+        renderHandler.post(() -> {
+            surfaceFuture.set(decodeSurface);
+        });
+        try {
+            return surfaceFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+            Log.e(TAG, "Could not set surface texture", e);
+            return null;
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
+    public void awaitNewImage() {
+        renderHandler.post(() -> {
+            final int TIMEOUT_MS = 500;
+
+            synchronized(frameSyncObject) {
+                while (!frameAvailable) {
+                    try {
+                        frameSyncObject.wait(TIMEOUT_MS);
+                        if (!frameAvailable) {
+                            throw new RuntimeException("Surface frame wait timed out");
+                        }
+                    } catch (InterruptedException e) {
+                        Log.e(TAG, "Failed to wait for available frame");
+                    }
+                }
+                frameAvailable = false;
+            }
+
+            // Latch the data
+            surfaceTexture.updateTexImage();
+        });
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
+    public void drawImage() {
+        renderHandler.post(() -> {
+            textureRender.drawFrame(surfaceTexture);
+            EGL14.eglSwapBuffers(eglDisplay, eglSurface);
+        });
+    }
+
+    @Override
+    public void onFrameAvailable(SurfaceTexture surfaceTexture) {
+        Log.d(TAG, "New frame available");
+        synchronized (frameSyncObject) {
+            if (frameAvailable) {
```

Comment:
any reason why you don't directly call drawImage here?

Reply:
Calling the method directly in onFrameAvaialble doesn't work, but I haven't figured out why yet. I'll make a note to look into this more in a separate CL.